CMakeLists.txt

```diff
@@ -0,0 +1,284 @@
+#          and their dependents.
+#  - 3.13: The target_link_libraries() command may now be called to modify targets created
+#          outside the current directory.
+#  - 3.14: The FindSQLite3 module was added to find the SQLite v3.x library.
+#  - 3.15: A new imported target Boost::headers is now defined (same as Boost::boost).
+#  - 3.18: The CheckLinkerFlag module has been added to provide a facility to check validity
+#          of link flags.
+cmake_minimum_required(VERSION 3.10)
+
+project("Bitcoin Core"
+  VERSION 23.0.0
+  DESCRIPTION "Bitcoin client software"
+  LANGUAGES CXX C ASM
+)
+
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake/modules)
+
+set(PACKAGE_NAME ${PROJECT_NAME})
+set(PACKAGE_VERSION ${PROJECT_VERSION})
+set(CLIENT_VERSION_MAJOR ${PROJECT_VERSION_MAJOR})
+set(CLIENT_VERSION_MINOR ${PROJECT_VERSION_MINOR})
+set(CLIENT_VERSION_BUILD ${PROJECT_VERSION_PATCH})
+set(CLIENT_VERSION_IS_RELEASE "false")
+set(PACKAGE_BUGREPORT "https://github.com/bitcoin/bitcoin/issues")
+set(PACKAGE_URL "https://bitcoincore.org/")
+set(COPYRIGHT_YEAR "2022")
+set(COPYRIGHT_HOLDERS "The %s developers")
+set(COPYRIGHT_HOLDERS_SUBSTITUTION ${PROJECT_NAME})
+set(COPYRIGHT_HOLDERS_FINAL "The Bitcoin Core developers")
+
+if(MSVC)
+  set(CMAKE_CXX_STANDARD 20)
+else()
+  set(CMAKE_CXX_STANDARD 17)
+endif()
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
+set(CMAKE_CXX_EXTENSIONS OFF)
+include(cmake/check_filesystem.cmake)
+
+# Configurable options.
+option(BUILD_DAEMON "Build bitcoind" ON)
+option(BUILD_CLI "Build bitcoin-cli" ON)
+option(BUILD_TX "Build bitcoin-tx" ON)
+option(BUILD_UTIL "Build bitcoin-util" ON)
+option(BUILD_BITCOINCONSENSUS_LIB "Build bitcoinconsensus shared library" ON)
+option(BUILD_BITCOINKERNEL_LIB "Build experimental bitcoinkernel shared library" ON)
+option(ENABLE_WALLET "Enable wallet" ON)
+option(BUILD_WALLET_TOOL "Build bitcoin-wallet tool" ${ENABLE_WALLET})
+option(REDUCE_EXPORTS "Attempt to reduce exported symbols in the resulting executables" OFF)
+
+include(cmake/try_append_cxxflag.cmake)
+# Define custom "RelWithAsserts" build type
+set(release_cxx_flags "")
+try_append_cxxflag(release_cxx_flags "-pipe")
+try_append_cxxflag(release_cxx_flags "-O2")
+try_append_cxxflag(release_cxx_flags "-g")
+if(CMAKE_CXX_COMPILER_ID STREQUAL GNU)
+  # All versions of gcc that we commonly use for building are subject to bug
+  # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90348. To work around that, set
+  # -fstack-reuse=none for all gcc builds.
+  try_append_cxxflag(release_cxx_flags "-fstack-reuse=none")
+endif()
+set(CMAKE_CXX_FLAGS_RELWITHASSERTS "${release_cxx_flags}" CACHE STRING
+  "Flags used by the C++ compiler during \"RelWithAsserts\" builds."
+  FORCE
+)
+# Define custom "RelHardened" build type
+set(hardened_cxx_flags "")
+try_append_cxxflag(hardened_cxx_flags "-Wstack-protector")
+try_append_cxxflag(hardened_cxx_flags "-fstack-protector-all")
```

Comment:
This test fails when cross-compiling for Windows, and we end up building without `-fstack-protector-all`.
```bash
/usr/lib/gcc/x86_64-w64-mingw32/11.2.1/../../../../x86_64-w64-mingw32/bin/ld: CMakeFiles/cmTC_c6384.dir/objects.a(src.cxx.obj):src.cxx:(.text+0x35): undefined reference to `__stack_chk_fail'
/usr/lib/gcc/x86_64-w64-mingw32/11.2.1/../../../../x86_64-w64-mingw32/bin/ld: CMakeFiles/cmTC_c6384.dir/objects.a(src.cxx.obj):src.cxx:(.rdata$.refptr.__stack_chk_guard[.refptr.__stack_chk_guard]+0x0): undefined reference to `__stack_chk_guard'
```

Reply:
Thanks! [Fixed](https://github.com/bitcoin/bitcoin/pull/25797#issuecomment-1218798376).